module/zfs/vdev.c

```diff
@@ -3466,6 +3453,12 @@ vdev_set_state(vdev_t *vd, boolean_t isopen, vdev_state_t state, vdev_aux_t aux)
 		vd->vdev_removed = B_FALSE;
 	}
 
+	/*
+	 * Notify ZED of any state change to a leaf vdev
+	 */
+	if (vd->vdev_ops->vdev_op_leaf)
+		zfs_post_state_change(spa, vd, save_state);
```

Comment:
The save_state value (aka last state) is transient and only exists in the context of `vdev_set_state()` function. I intentionally didn't carve out a new vdev instance variable slot and since `zfs_post_state_change()` is only called from this funtion it seemed fine to adjust the prototype.  I agree that there could be confusion and suggest that we call the prototype arg `last_state` and the corresponding payload key.

At some point we should shore up the vdev state logic. There are gaps where, for example, the degraded state of a vdev is lost after the reopening.  I'd like to propose using phase 2 to tackle that effort.  The next commit here of  `zfs_post_state_change` should expose all vdev state changes to/from healthy and non-healthy.